x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SLMGetExpiredSnapshotsAction.java

```diff
@@ -0,0 +1,304 @@
+                                    l.onResponse(null);
+                                }))
+                        )
+                    );
+                }
+            }
+        }
+    }
+
+    private static class SnapshotDetailsByPolicy {
+        private final Map<String, Map<SnapshotId, RepositoryData.SnapshotDetails>> snapshotsByPolicy = new HashMap<>();
+
+        synchronized void add(SnapshotId snapshotId, RepositoryData.SnapshotDetails snapshotDetails) {
+            assert RETAINABLE_STATES.contains(snapshotDetails.getSnapshotState());
+            final var slmPolicy = snapshotDetails.getSlmPolicy();
+            if (Strings.hasLength(slmPolicy)) {
+                snapshotsByPolicy.computeIfAbsent(slmPolicy, ignored -> new HashMap<>()).put(snapshotId, snapshotDetails);
+            }
+        }
+
+        <T> Stream<T> flatMap(BiFunction<String, Map<SnapshotId, RepositoryData.SnapshotDetails>, Stream<T>> fn) {
+            // no need for synchronized here, building the map entirely happens-before reading it
+            return snapshotsByPolicy.entrySet().stream().flatMap(entry -> fn.apply(entry.getKey(), entry.getValue()));
+        }
+    }
+
+    private static void getSnapshotDetailsByPolicy(
+        Repository repository,
+        RepositoryData repositoryData,
+        ActionListener<SnapshotDetailsByPolicy> listener
+    ) {
+        final var snapshotDetailsByPolicy = new SnapshotDetailsByPolicy();
+        final var snapshotsWithMissingDetails = new ArrayList<SnapshotId>();
+
+        for (final var snapshotId : repositoryData.getSnapshotIds()) {
+            if (repositoryData.hasMissingDetails(snapshotId)) {
+                snapshotsWithMissingDetails.add(snapshotId);
+            } else {
+                snapshotDetailsByPolicy.add(snapshotId, Objects.requireNonNull(repositoryData.getSnapshotDetails(snapshotId)));
+            }
+        }
+
+        if (snapshotsWithMissingDetails.isEmpty()) {
+            listener.onResponse(snapshotDetailsByPolicy);
+        } else {
+            // rare bwc path, older repositories might not have snapshot details fully populated yet
+            logger.debug(
+                "[{}]: retrieving snapshot details from snapshot info for {}",
+                repository.getMetadata().name(),
+                snapshotsWithMissingDetails
+            );
+            repository.getSnapshotInfo(
+                new GetSnapshotInfoContext(
+                    snapshotsWithMissingDetails,
+                    false,
+                    () -> false,
+                    (ignored, snapshotInfo) -> snapshotDetailsByPolicy.add(
+                        snapshotInfo.snapshotId(),
+                        RepositoryData.SnapshotDetails.fromSnapshotInfo(snapshotInfo)
+                    ),
+                    listener.map(ignored -> snapshotDetailsByPolicy)
+                )
+            );
+        }
+    }
+
+    private static List<Tuple<SnapshotId, String>> getSnapshotsToDelete(
```

Comment:
Thanks for all the details in this function ❤️ It's a real pleasure to read. 

Would we benefit from a bit more targeted testing for this function if it was package-visible and target every if/else branch with unit tests? (similar to the previous `SnapshotRetentionTask#snapshotEligibleForDeletion`)

Reply:
I'm not sure this adds much (the existing test already covers all the branches) but I added some more focussed tests in eef1e63b42609a8c5612d665a03b2520c091f774.